### ISSUE TITLE
method of specifying a libname in ccall may cause segfault

### ISSUE THREAD

**Issue by pixel70** [JuliaLang/julia#29602]
In v0.6.4 the lines 
```
lib = "./libtest.so"
v = ccall((:test,lib),Int32,())
```
can be successfully used to call a simple function named `test` in a shared library named `libtest.so`

But in v1.0.1 they lead to a segfault that terminates the REPL.

A simple workaround is to change the line to
```
  v = ccall((:test,"./libtest.so"),Int32,())
```
which works successfully in both v0.6 and v1.0

I suggest adding error checks to the ccall function to catch tuples that are not of the form `(:funcname,"libname")` in order to avoid segfaulting.

The segfaulting can be demonstrated using a C file named `test.c` that contains
```
#include <stdio.h>
int test( )
{ 
 printf(“test\n”);
 return 999;
}
```
compiled with:
```
gcc -shared -fPIC test.c -o libtest.so
```
and a julia file `m.jl` that contains:
```
function test()
  lib = “./libtest.so”
  v = ccall((:test,lib),Int32,())
  println(v)
end

println(“hello”)
test()
println(“ok”)
```

**Comment by RalphAS** [comment]
(in hope of routing this to the appropriate experts)
IIUC the segfault occurs in codegen before the usual sort of argument check is done:
```julia
signal (11): Segmentation fault
in expression starting at no file:0
LoadInst at /scratch/build/julia-1.x-local/usr/include/llvm/IR/Instructions.h:212 [inlined]
CreateLoad at /scratch/build/julia-1.x-local/usr/include/llvm/IR/IRBuilder.h:1178
emit_varinfo at /scratch/build/julia/src/codegen.cpp:3377
emit_expr at /scratch/build/julia/src/codegen.cpp:3425
emit_call at /scratch/build/julia/src/codegen.cpp:3114
emit_expr at /scratch/build/julia/src/codegen.cpp:3872
interpret_symbol_arg at /scratch/build/julia/src/ccall.cpp:656
emit_ccall at /scratch/build/julia/src/ccall.cpp:1470
emit_expr at /scratch/build/julia/src/codegen.cpp:3882
emit_ssaval_assign at /scratch/build/julia/src/codegen.cpp:3585
emit_stmtpos at /scratch/build/julia/src/codegen.cpp:3771 [inlined]
emit_function at /scratch/build/julia/src/codegen.cpp:6203
jl_compile_linfo at /scratch/build/julia/src/codegen.cpp:1141
...
```

**Comment by maleadt** [comment]
Doesn't actually need a C file to exist:

```
function foo(x)
    lib = "whatever"
    ccall((:whatever, lib), Nothing, ())
end
foo(nothing)
```

Triggers an assertion:

```
julia: /home/tim/Julia/julia-dev/src/julia.h:806: jl_value_t *jl_array_ptr_ref(void *, size_t): Assertion `i < jl_array_len(a)' failed.

signal (6): Aborted
in expression starting at /home/tim/Julia/julia-dev/wip.jl:5
gsignal at /usr/lib/libc.so.6 (unknown line)
abort at /usr/lib/libc.so.6 (unknown line)
__assert_fail_base.cold.0 at /usr/lib/libc.so.6 (unknown line)
__assert_fail at /usr/lib/libc.so.6 (unknown line)
jl_array_ptr_ref at /home/tim/Julia/julia-dev/src/julia.h:806 [inlined]
slot_symbol at /home/tim/Julia/julia-dev/src/codegen.cpp:729 [inlined]
emit_local at /home/tim/Julia/julia-dev/src/codegen.cpp:3494 [inlined]
emit_expr at /home/tim/Julia/julia-dev/src/codegen.cpp:3867
emit_call at /home/tim/Julia/julia-dev/src/codegen.cpp:3191
emit_expr at /home/tim/Julia/julia-dev/src/codegen.cpp:3957
interpret_symbol_arg at /home/tim/Julia/julia-dev/src/ccall.cpp:656
emit_ccall at /home/tim/Julia/julia-dev/src/ccall.cpp:1475
```

**Comment by Andrey1994** [comment]
Any news here? I have this issue as well, it makes it impossible to write a function which returns libname based on OS and use it in ccall

I have this problem with 1.4.0 installed from src and with latest release for fedora 31(1.2.0)

**Comment by matthijscox** [comment]
FYI, the common pattern seems to be to use a global `const` in your package/module for library names. This does not cause this issue:
```
module MyModule
    const lib = "whatever"
    function foo(x)
        ccall((:whatever, lib), Nothing, ())
    end
end
using .Module
MyModule.foo(nothing)
```

Returns an expected error
```
ERROR: could not load library "whatever"
The specified module could not be found. 
```

**Comment by StefanKarpinski** [comment]
Non-constant libnames should have been a compilation error for earlier Julia versions, but they are allowed in Julia 1.6 with the caveat that if you change the value after a ccall is compiled, it may not be reflected in the library that is actually called. cc @staticfloat, @JeffBezanson if you could remind me what actually changed here.

**Comment by staticfloat** [comment]
Here's a demonstration:

```
$ cat lib*.c
// liba.c
#include <stdio.h>

void foo() {
    printf("Hello from liba!\n");
}
// libb.c
#include <stdio.h>

void foo() {
    printf("Hello from libb!\n");
}
$ gcc -o liba.dylib -shared liba.c
$ gcc -o libb.dylib -shared libb.c
```

Then, in Julia 1.5:
```
julia> gen_libname() = rand() > 0.5 ? "liba" : "libb"
       ccall((:foo, gen_libname()), Cvoid, ())
ERROR: TypeError: in ccall: first argument not a pointer or valid constant expression, expected Ptr, got a value of type Tuple{Symbol,String}
Stacktrace:
 [1] top-level scope at ./REPL[1]:2
```

Whereas in Julia 1.6:
```
julia> gen_libname() = rand() > 0.5 ? "liba" : "libb"
       ccall((:foo, gen_libname()), Cvoid, ())
Hello from libb!

julia> gen_libname() = rand() > 0.5 ? "liba" : "libb"
       ccall((:foo, gen_libname()), Cvoid, ())
Hello from liba!

julia> gen_libname() = rand() > 0.5 ? "liba" : "libb"
       ccall((:foo, gen_libname()), Cvoid, ())
Hello from libb!

julia> gen_libname() = rand() > 0.5 ? "liba" : "libb"
       ccall((:foo, gen_libname()), Cvoid, ())
Hello from libb!
```

Notice, however, that since these are top-level statements and the `ccall()` is being compiled again and again, each `ccall()` is different.  If we were to run this in a loop where the `ccall()` is being compiled just once:

```
julia> gen_libname() = rand() > 0.5 ? "liba" : "libb"
       global libname
       function call_foo(N)
           for idx in 1:N
               global libname = gen_libname()
               @info("I think I'm about to call $(libname)")
               ccall((:foo, libname), Cvoid, ())
           end
       end
       call_foo(10)
[ Info: I think I'm about to call libb
Hello from libb!
[ Info: I think I'm about to call libb
Hello from libb!
[ Info: I think I'm about to call libb
Hello from libb!
[ Info: I think I'm about to call libb
Hello from libb!
[ Info: I think I'm about to call liba
Hello from libb!
[ Info: I think I'm about to call liba
Hello from libb!
[ Info: I think I'm about to call libb
Hello from libb!
[ Info: I think I'm about to call libb
Hello from libb!
[ Info: I think I'm about to call libb
Hello from libb!
[ Info: I think I'm about to call libb
Hello from libb!
```

I had to make `libname` a global to defeat the check that `ccall()` has about referencing local variable names.  The reason why it always says `"Hello from libb!"` is that the lookup within the given library is done not at runtime, but at compile-time.  This means that whatever library you pass in to this `ccall()` the first time is the one that will be used by that `ccall()` from then on.  This is an optimization to avoid needlessly looking up symbol addresses again and again.

If you need the flexibility to change library (or function name) dynamically, use `Libdl.dlsym()` to look up a function address, then pass that directly to `ccall`.

**Comment by matthijscox** [comment]
I've been wondering if I'd rather have the error, instead of this behavior. Imagine debugging this behavior when you are unaware of it.

In order to dynamically switch library in v1.5, what I did now is hardcode some cglobals and then dispatch on those. In your example it would be like this:

```
const LIBA = "path/to/liba.dll"
const LIBB = "path/to/libb.dll"

abstract type Library end
struct LibA <: Library end
struct LibB <: Library end

foo_cglobal(::LibA) = cglobal((:foo, LIBA)) 
foo_cglobal(::LibB) = cglobal((:foo, LIBB)) 

call_foo(lib::Library) = ccall(foo_cglobal(lib), Cvoid, ())
```

I even started to use @eval to generate the cglobals. Though perhaps I should look into CBindingGen.jl

**Comment by StefanKarpinski** [comment]
This feature is needed to massively speed up the loading of JLLs in 1.6. Even checking if the binding has changed is too slow to be done on every ccall. The compiler team had a long discussion about this and there wasn't really any better solution that we could come up with. It's true that it may be quite hard to debug, however.

**Comment by vtjnash** [comment]
Fixed in v1.6 (or earlier?)
```
julia> function foo(x)
           lib = "whatever"
           ccall((:whatever, lib), Nothing, ())
       end
ERROR: syntax: ccall function name and library expression cannot reference local variables
Stacktrace:
 [1] top-level scope
   @ REPL[1]:1
```
